### ISSUE TITLE
Query string does not honor required/not required

### ISSUE THREAD

**Issue by thomastoye** [serverless/serverless#5034]
# This is a Bug Report

## Description

* What went wrong?

I created a query string parameter set to `true`, which marks it as required. However, nothing happens when it's not passed.

* What did you expect should have happened?

I expect to see an error when the parameter is missing.

* What was the config you used?

See below for `serverless.yml`.

* What stacktrace or error message from your provider did you see?

None. Expected to see an error when query string parameter is missing.

---

I want my request to have a certain query string parameter. I saw the following [here](https://serverless.com/framework/docs/providers/aws/events/apigateway/#request-parameters):

>To pass optional and required parameters to your functions, so you can use them in API Gateway tests and SDK generation, marking them as true will make them required, false will make them optional.

However, when marking a query string as `true` does not seem to make it required. When using the following code (clean start from `serverless create --template aws-nodejs-typescript`), all requests pass to the `hello` path, even if the query string is not set.

`serverless.yml`:

```
service:
  name: example-test

plugins:
  - serverless-webpack
  - serverless-offline

provider:
  name: aws
  runtime: nodejs6.10

functions:
  hello:
    handler: handler.hello
    events:
      - http:
          method: get
          path: hello
          request:
            parameters:
              querystrings:
                url: true
```

Similar or dependent issues:
* https://stackoverflow.com/questions/49065665/querystring-parameters-are-not-getting-validated-even-after-setting-the-params-t

## Additional Data

* ***Serverless Framework Version you're using***: 1.27.3
* ***Operating System***: Ubuntu 18.04 LTS
* ***Stack Trace***: N/A
* ***Provider Error messages***: N/A

**Comment by ctindel** [comment]
I think this is related to the fact that there's no way right now to enable the request validator for body/query params right?

https://github.com/serverless/serverless/issues/3464

**Comment by Rigellute** [comment]
Also seeing this behaviour 
```
Framework Core: 1.60.5
Plugin: 3.2.7
SDK: 2.2.1
Components Core: 1.1.1
Components CLI: 1.4.0
```

**Comment by akleiber** [comment]
Querystring paramters are marked as required, but the missing part is the `RequestValidator`.

![Bildschirmfoto 2020-02-04 um 08 55 24](https://user-images.githubusercontent.com/1506973/73724707-3c0f7300-472c-11ea-8b1b-c9ef3b8e426c.png)

**Comment by akleiber** [comment]
If you set params as required via serverless syntax you then can just also activate the validator for parameters like so

```
Resources:
  ParameterRequestValidator:
    Type: AWS::ApiGateway::RequestValidator
    Properties:
      Name: ParameterRequestValidator
      RestApiId:
        Ref: ApiGatewayRestApi
      ValidateRequestBody: false
      ValidateRequestParameters: true

  ApiGatewayMethodNameOfYourApiLookItUpInYourTemplatet:
    Properties:
      RequestValidatorId:
        Ref: ParameterRequestValidator
```

**Comment by Rigellute** [comment]
@akleiber thank you 🙏

**Comment by Pitasi** [comment]
I'm struggling with this too.

What's "ApiGatewayMethodNameOfYourApiLookItUpInYourTemplate" you're referring to, @akleiber? 
I tried a bunch of different combinations with no luck :)

cc @Rigellute maybe you can anser too

**Comment by Pitasi** [comment]
Oookay I got it!
It should be: `ApiGatewayMethodFooGet`, where `foo` is my handler name and `get` is the HTTP method I'm handling.

**Comment by yash-awsops** [comment]
Here is the working template:

In serverless.yml add this

       plugins:
           - serverless-reqvalidator-plugin
       functions:
           events:
              -http:
                path: hello
                method: get
                reqValidatorName: ParameterRequestValidator
        resources:

         Resources:
             ParameterRequestValidator:
                Type: AWS::ApiGateway::RequestValidator
                Properties:
                       Name: ParameterRequestValidator
                       RestApiId:
                             Ref: ApiGatewayRestApi
                       ValidateRequestBody: false
                       ValidateRequestParameters: true

**Comment by shadowdogg** [comment]
> 
> 
> Oookay I got it!
> It should be: `ApiGatewayMethodFooGet`, where `foo` is my handler name and `get` is the HTTP method I'm handling.

Don't suppose you could be a bit more specific for people like me please.

This is what I got from your description:

```
  list:
    # Defines an HTTP API endpoint that calls the main function in list.js
    # - path: url path is /deals
    # - method: GET request
    handler: list.main
    events:
      - http:
          path: deals
          method: get
          cors: true
          authorizer: aws_iam
          request:
            parameters:
              querystrings:
                country: true
                type: true
                categoryId: true
                yearMonth: true
                batchLimit: true
    Properties:
      RequestValidatorId:
        Ref: ParameterRequestValidator
```

**Comment by Pitasi** [comment]
@shadowdogg iirc you just need to add this at the bottom of your `serverless.yml`:

```
Resources:
  ParameterRequestValidator:
    Type: AWS::ApiGateway::RequestValidator
    Properties:
      Name: ParameterRequestValidator
      RestApiId:
        Ref: ApiGatewayRestApi
      ValidateRequestBody: false
      ValidateRequestParameters: true

  ApiGatewayMethodListGet:
    Properties:
      RequestValidatorId:
        Ref: ParameterRequestValidator
```

(note how I used ApiGatewayMethod**ListGet** since your handlest it's called `list` and the method is `get`)

**Comment by shadowdogg** [comment]
> 
> 
> @shadowdogg iirc you just need to add this at the bottom of your `serverless.yml`:
> 
> ```
> Resources:
>   ParameterRequestValidator:
>     Type: AWS::ApiGateway::RequestValidator
>     Properties:
>       Name: ParameterRequestValidator
>       RestApiId:
>         Ref: ApiGatewayRestApi
>       ValidateRequestBody: false
>       ValidateRequestParameters: true
> 
>   ApiGatewayMethodListGet:
>     Properties:
>       RequestValidatorId:
>         Ref: ParameterRequestValidator
> ```
> 
> (note how I used ApiGatewayMethod**ListGet** since your handlest it's called `list` and the method is `get`)

Yeah, still doesn't work for me

The CloudFormation template is invalid: Template format error: [/Resources/ApiGatewayMethodListGet] Every Resources object must contain a Type member.

**Comment by shadowdogg** [comment]
For those of you failing to see this, like I also did.

This is what you need to do in plain english.

Turn

ApiGatewayMethodNameOfYourApiLookItUpInYourTemplate

to 

APIGatewayMethod<1><2>

[![API Gateway][1]][1]

In my case, it was APIGatewayDealsGet

The thing I was looking at was my handler name in serverless

       list:
        # Defines an HTTP API endpoint that calls the main function in list.js
        # - path: url path is /deals
        # - method: GET request
        handler: list.main
        events:
          - http:
              path: deals
              method: get
              cors: true
              authorizer: aws_iam
              request:
                parameters:
                  querystrings:
                    country: true
                    type: true

  [1]: https://i.stack.imgur.com/Hnrcr.png

Alternatively, if this does not work, check the s3 bucket, mine was called xxxxxxx-ap-serverlessdeploymentbuck-1epdp60eqveqr and go to serverless > yyyyyyyyyyy > aaaa > 
timestamp > compiled-cloudformation-template.json

And look for the name of your method in there, example mine was:

		"ApiGatewayMethodDealsGet": {
			"Type": "AWS::ApiGateway::Method",
			"Properties": {
				"HttpMethod": "GET",
				"RequestParameters": {

**Comment by shwetajoshi601** [comment]
Is there any update on this? I am facing the same issue.

**Comment by shadowdogg** [comment]
> 
> 
> Is there any update on this? I am facing the same issue.

I literally gave a full explaination right above your comment

**Comment by mamapi** [comment]
@shadowdogg what type should be used for that resource? 
Now I'm getting "The CloudFormation template is invalid: Template format error ....  Every Resources object must contain a Type member."

I'm reffering to 
` ApiGatewayMethodListGet:
    Properties:
      RequestValidatorId:
        Ref: ParameterRequestValidator`

**Comment by bmcilw1** [comment]
This did not solve my issue unfortunately :( 

https://github.com/serverless/serverless/issues/5034#issuecomment-581832806

**Comment by flexjames** [comment]
> If you set params as required via serverless syntax you then can just also activate the validator for parameters like so
> 
> ```
> Resources:
>   ParameterRequestValidator:
>     Type: AWS::ApiGateway::RequestValidator
>     Properties:
>       Name: ParameterRequestValidator
>       RestApiId:
>         Ref: ApiGatewayRestApi
>       ValidateRequestBody: false
>       ValidateRequestParameters: true
> 
>   ApiGatewayMethodNameOfYourApiLookItUpInYourTemplate:
>     Properties:
>       RequestValidatorId:
>         Ref: ParameterRequestValidator
> ```

Thanks this worked.  Quick question though, does this method work for path parameters?  Also, if I have a nested route, how would that work?  So if my route is admin/someRoute, what would the ApiGatewayMethod<name><requestType> look like?  I tried ApiGatewayMethodAdminSomerouteGet and it didn't work?

**Comment by Shahid1993** [comment]
> > If you set params as required via serverless syntax you then can just also activate the validator for parameters like so
> > ```
> > Resources:
> >   ParameterRequestValidator:
> >     Type: AWS::ApiGateway::RequestValidator
> >     Properties:
> >       Name: ParameterRequestValidator
> >       RestApiId:
> >         Ref: ApiGatewayRestApi
> >       ValidateRequestBody: false
> >       ValidateRequestParameters: true
> > 
> >   ApiGatewayMethodNameOfYourApiLookItUpInYourTemplate:
> >     Properties:
> >       RequestValidatorId:
> >         Ref: ParameterRequestValidator
> > ```
> 
> Thanks this worked. Quick question though, does this method work for path parameters? Also, if I have a nested route, how would that work? So if my route is admin/someRoute, what would the ApiGatewayMethod look like? I tried ApiGatewayMethodAdminSomerouteGet and it didn't work?

This doesn't work for path parameters.

To get the APIGatewayMethod name for your nested route, you can look in the `compiled-cloudformation-template.json` file inside your serverless deployment bucket.
